### DISCUSSION TITLE
MongoDB update nested documents

### DISCUSSION THREAD

**Discussion by lucafaggianelli** [prisma/prisma#16467] [Q&A] [answered]
Hey there,

I have a data model to represent a CO2 emissions report that looks like this in MongoDB:

```js
{
  electricalEnergy: {
    value: 1000,
    co2Emissions: {
      catA: 100,
      catB: 50
    }
  },
  transportationMeans: [
    {
      value: { type: 'car', engine: 'LPG', consumption: 1000 },
      co2Emissions: { catA: 20, catB: 3 }
    }
  ],
  anotherOne:
    {
      value: { obj: 'prop' },
      co2Emissions: { catA: 20, catB: 3 }
    }
}
```

On the frontend I have a form to fill the report data and the problem is that to update the model on the DB, I need to create a Mongo query with the `set` operator, I'm using this code:

```js
data = Object.fromEntries(
      Object.entries(data)
        .map(([key, value]) => [key, { set: value }])
    )
```

to obtain this result:

```js
db.reportData.update({where, data: {
    electricalEnergy: {
      set: {
        value: 1000
      }
    }
}})
```
I don't like it much, but it works. 

### The problem
Though the problem is that the main document has around 100 fields and when I run the query, I receive an Error from MongoDB Atlas: `Command failed (AtlasError): Pipeline length greater than 50 not supported)`.

Now the question is: can I avoid adding the `set` operator to all my document's fields? Maybe just using a `set` for the entire doc?

Moreover, do you advise a different data model? I'm open to using a SQL DB but I don't know how to handle a `value` field that can be a number or an object...maybe using PostgreSQL JSON format?

### Prisma schema
Talking about the Prisma schema, my report model is a document where each property is the name of the CO2 contribution (energy, cars, etc.) and the value is a nested document with the co2Emissions and value (a number) properties. Things get complicated as in some cases, the value is an object and sometimes there are multiple values (like in `transportationMeans`) so I needed to create ad-hoc types just for that field:

```prisma
type TransportationMeansValueType {
  type       String
  engine String
  standardEuro  String
  size        String
  consumption       Float
}

type TransportationMeansType {
  value       MezziTrasportoValueType
  co2         Co2Emissions?
  notes       String?
  attachments String[]                @default([]) @db.ObjectId
}

// Generic type, most of the time I use this
type ReportDataField {
  value       Float
  co2         Co2Emissions?
  notes       String?
  attachments String[]      @default([]) @db.ObjectId
}

model ReportData {
  electricalEnergy  ReportDataField
  transportationMeans  TransportationMeansType[]
}
```

Thanks so much, any advise would be super appreciated!
Luca

**Comment by nurul3101** [comment]
Hello @lucafaggianelli 👋 

Based on the error message that you have shared, did you try out the solution mentioned in this Stack Overflow [Answer](https://stackoverflow.com/questions/63754063/mongodb-atlas-pipeline-length-greater-than-50-not-supported)?
They recommend splitting your pipeline into chunks and using [$facet](https://www.mongodb.com/docs/manual/reference/operator/aggregation/facet/) operator.

> Maybe just using a set for the entire doc

This seems like a viable option as well, this should reduce the number of pipeline operations.
Compared to splitting pipeline into chunks, using `set` for entire doc seems like a more straightforward approach.

Let me know if you have any other queries.

**Reply by pantharshit00** [reply]
👋 Hello @lucafaggianelli,  did you perhaps have a chance to check @nurul3101's recommendation? Let us know if you still have any questions!
 
Or, if you've found a solution, we'd love to know how you solved this problem. Feel free to post it in this thread and "Mark as answer" if this discussion is now solved!

**Reply by lucafaggianelli** [reply]
Hey, sorry for the late reply and thanks for your suggestions, though I was able to solve the problem by cutting the query in smaller queries.

My big question to you is why Prisma is generating an aggregation query when I run this:
```ts
prisma.myModel.update({})
```

In the log I see this, basically the prisma query is transformed into a list of $set operations:
```
prisma:query db.ReportData.updateMany({ _id: { $in: [ ObjectId("63cd4e4e4c3a96b5c7adbd1e"), ], }, },[{ $set: { updatedAt: { $literal: DateTime("2023-01-22 17:03:28.814 +00:00:00"), }, }, },
[dx:next         ] { $set: { lettoriOttici: { $literal: null, }, }, },
[dx:next         ] { $set: { mascherine: { $literal: null, }, }, },
[dx:next         ] { $set: { mezziTrasporto: { $literal: [ ], }, }, },
[dx:next         ] { $set: { monitor: { $literal: null, }, }, },
[dx:next         ] { $set: { monitorUfficio: { $literal: null, }, }, },
[dx:next         ] { $set: { olioEsausto: { $literal: null, }, }, },
[dx:next         ] { $set: { ombrelloni: { $literal: null, }, }, },
[dx:next         ] { $set: { pantaloniLavoratori: { $literal: null, }, }, },
[dx:next         ] { $set: { rabboccoGasRefrigerante: { $literal: null, }, }, },
[dx:next         ] { $set: { rifiutiElettroniciEORifiutiIngombrantiARecupero: { $literal: null, }, }, },
[dx:next         ] { $set: { rifiutoIndifferenziato: { $literal: null, }, }, },
[dx:next         ] { $set: { scarpeLavoratori: { $literal: null, }, }, },
}])
```

**Comment by lucafaggianelli** [comment, marked as answer]
At the end I solved my issue cutting the original data object into smaller objects:

```ts
const reportDataEntries = Object.entries(bigObjectWithMoreThan50Keys));
const reportData1 = Object.fromEntries(reportDataEntries.slice(0, 45));
const reportData2 = Object.fromEntries(reportDataEntries.slice(45));

await prisma.$transaction([
  prisma.reportData.update({
    data: reportData1,
    where: { id: report.data.id },
  }),

  prisma.reportData.update({
    data: reportData2,
    where: { id: report.data.id },
  }),
]);
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
